addons/udes_stock/models/stock_picking.py

```diff
@@ -2753,3 +2753,46 @@ def _get_pick_details(self):
             "submitted_value": self.origin,
         }
         return pick_details
+
+    @api.model
+    def _set_original_picking_on_historical_pickings(self):
+        """
+        Set the Original Picking for historical pickings that do not have an Original Picking set.
+
+        Carries out update as superuser to avoid any permissions issues, as the method is simply
+        setting Original Picking to the correct value, and doesn't take any input variables.
+        """
+        Picking = self.env["stock.picking"]
+
+        pickings_to_update = Picking.sudo().with_context(prefetch_fields=False).search(
+            [("u_original_picking_id", "=", False)], order="id"
+        )
+
+        picking_ids = pickings_to_update.ids
+        picking_count = len(picking_ids)
+
+        if picking_count:
+            _logger.info("Setting Original Picking value on %s pickings", picking_count)
+
+            update_datetime = fields.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S.%f")
+
+            update_query = """
+            UPDATE stock_picking
+            SET
+            write_uid = {user_id},
+            write_date = '{update_datetime}',
+            u_original_picking_id =
+            CASE
+            WHEN backorder_id IS NOT null THEN backorder_id
```

Comment:
What happens in the recursive hole? Pick 3 backorder of pick 2 backorder of pick 1?

Reply:
Bad miss from me. Good spot. :)

Now using a recursive CTE so I can get the picking that is top of the tree, for those pickings that are backorders. Tested against the value set against newly created pickings on the system.

Timed at 1 minute 45 seconds for a database with roughly 800k pickings but open to suggestions on performance if any, the query can take a little while to get used to at first.